src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs

```diff
@@ -0,0 +1,181 @@
+            {
+                return null;
+            }
+
+            int patch = 0;
+            int patchStart = minorSeparator + 1;
+            int patchSeparator = semanticVersionString.FindFirstNotOf("0123456789", patchStart);
+            if (patchSeparator == -1)
+            {
+                if (!int.TryParse(semanticVersionString.Substring(patchStart), out patch))
+                {
+                    return null;
+                }
+
+                return new SemanticVersion(major, minor, patch);
+            }
+
+            if (!int.TryParse(semanticVersionString.Substring(patchStart, patchSeparator - patchStart), out patch))
+            {
+                return null;
+            }
+
+            int preStart = patchSeparator;
+            int preSeparator = semanticVersionString.IndexOf("+", preStart);
+            if (preSeparator == -1)
+            {
+                return new SemanticVersion(major, minor, patch, semanticVersionString.Substring(preStart));
+            }
+            else
+            {
+                int buildStart = preSeparator + 1;
+                return new SemanticVersion(
+                    major,
+                    minor,
+                    patch,
+                    semanticVersionString.Substring(preStart, preSeparator - preStart),
+                    semanticVersionString.Substring(buildStart));
+            }
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null)
+            {
+                return false;
+            }
+
+            var other = obj as SemanticVersion;
+            if (other == null)
+            {
+                return false;
+            }
+
+            return this == other;
+        }
+
+        public override int GetHashCode()
+        {
+            return Major.GetHashCode() ^
+                   Minor.GetHashCode() ^
+                   Patch.GetHashCode() ^
+                   Pre.GetHashCode() ^
+                   Build.GetHashCode();
+        }
+
+        private static int Compare(SemanticVersion s1, SemanticVersion s2)
+        {
+            if (s1.Major != s2.Major)
+            {
+                return s1.Major > s2.Major ? 1 : -1;
+            }
+
+            if (s1.Minor != s2.Minor)
+            {
+                return s1.Minor > s2.Minor ? 1 : -1;
+            }
+
+            if (s1.Patch != s2.Patch)
+            {
+                return s1.Patch > s2.Patch ? 1 : -1;
+            }
+
+            if (string.IsNullOrEmpty(s1.Pre) != string.IsNullOrEmpty(s2.Pre))
+            {
+                return string.IsNullOrEmpty(s1.Pre) ? 1 : -1;
+            }
+
+            int preCompare = string.Compare(s1.Pre, s2.Pre);
+            if (preCompare != 0)
+            {
+                return preCompare;
+            }
+
+            return string.Compare(s1.Build, s2.Build);
```

Comment:
I see now that this is in fact all ported straight from core-setup (which makes sense, if we compare differently than hostfxr, it will lead to strange behavior.) 

I think we should call it FXVersion and copy the  "// Note: This is not SemVer" from fx_ver and say that we need to match hostfxr's version comparison.